File: app/src/main/java/com/pomo/sync/diagnostics/DiagnosticExport.kt
package com.pomo.sync.diagnostics

import java.io.OutputStream

internal enum class EvidenceArea {
    STATE_TRANSITION,
    FRONTIER,
    DISPOSITION,
    ROUTE,
    COMMIT_ACK,
    LIFECYCLE_GENERATION,
    RETRY_BACKOFF,
    PROVIDER_PROTECTION,
    CHECKPOINT_REHYDRATION,
    MIGRATION,
    COMPATIBILITY,
    SAFE_MODE,
    PERFORMANCE,
}

internal data class DiagnosticEvent(
    val monotonicMillis: Long,
    val area: EvidenceArea,
    val event: String,
    val fields: Map<String, String>,
    val deviceRef: String? = null,
    val operationRef: String? = null,
)

internal class DiagnosticRecorder(private val capacity: Int = 2_048) {
    private val events = ArrayDeque<DiagnosticEvent>()

    @Synchronized
    fun record(event: DiagnosticEvent) {
        DiagnosticRedaction.requireSafe(event)
        while (events.size >= capacity) events.removeFirst()
        events.addLast(event)
    }

    @Synchronized
    fun snapshot(): List<DiagnosticEvent> = events.toList()
}

internal object DiagnosticRedaction {
    private val allowedFields =
        setOf(
            "from",
            "to",
            "count",
            "sequence",
            "generation",
            "durationMs",
            "batchSize",
            "attempt",
            "delayMs",
            "outcome",
            "reasonCode",
            "routeKind",
            "sourceKind",
            "formatVersion",
        )
    private val forbidden = Regex("(?i)(key|secret|token|credential|capability|recovery|profile|photo|tag|history|payload|content|member)")
    fun requireSafe(event: DiagnosticEvent) {
        require(event.event.length in 1..80 && !forbidden.containsMatchIn(event.event))
        require(event.fields.size <= 24)
        event.fields.forEach { (key, value) ->
            require(key in allowedFields && value.length <= 120 && !forbidden.containsMatchIn(value)) { "Diagnostic field is not export-safe" }
        }
    }
}

internal data class DiagnosticExportResult(
    val bytesWritten: Long,
    val eventsWritten: Int,
    val truncated: Boolean,
    val cancelled: Boolean,
)

internal object DiagnosticExporter {
    const val MAX_BYTES: Long = 10L * 1024L * 1024L

    fun export(
        events: Sequence<DiagnosticEvent>,
        output: OutputStream,
        cancelled: () -> Boolean = { false },
    ): DiagnosticExportResult {
        val deviceAliases = linkedMapOf<String, String>()
        val operationAliases = linkedMapOf<String, String>()
        var bytes = 0L
        var count = 0
        for (event in events) {
            if (cancelled()) {
                return DiagnosticExportResult(bytes, count, truncated = false, cancelled = true)
            }
            DiagnosticRedaction.requireSafe(event)
            val line = encode(event, deviceAliases, operationAliases).toByteArray(Charsets.UTF_8)
            if (bytes + line.size > MAX_BYTES) {
                return DiagnosticExportResult(bytes, count, truncated = true, cancelled = false)
            }
            output.write(line)
            bytes += line.size
            count += 1
        }
        return DiagnosticExportResult(bytes, count, truncated = false, cancelled = false)
    }

    private fun encode(
        event: DiagnosticEvent,
        devices: MutableMap<String, String>,
        operations: MutableMap<String, String>,
    ): String {
        val fields =
            event.fields.toSortedMap().entries.joinToString(",") { (key, value) ->
                "\"${escape(key)}\":\"${escape(value)}\""
            }
        val device = event.deviceRef?.let { devices.getOrPut(it) { "device-${devices.size + 1}" } }
        val operation = event.operationRef?.let { operations.getOrPut(it) { "operation-${operations.size + 1}" } }
        return buildString {
            append(
                "{\"v\":1,\"monotonicMillis\":${event.monotonicMillis}," +
                    "\"area\":\"${event.area.name}\",\"event\":\"${escape(event.event)}\"," +
                    "\"fields\":{$fields}",
            )
            if (device != null) append(",\"deviceAlias\":\"$device\"")
            if (operation != null) append(",\"operationAlias\":\"$operation\"")
            append("}\n")
        }
    }

    private fun escape(value: String): String =
        buildString {
            value.forEach { character ->
                when (character) {
                    '\\' -> append("\\\\")
                    '"' -> append("\\\"")
                    '\n' -> append("\\n")
                    '\r' -> append("\\r")
                    '\t' -> append("\\t")
                    else ->
                        if (character.code < 0x20) {
                            append("\\u%04x".format(character.code))
                        } else {
                            append(character)
                        }
                }
            }
        }
}

File: extension/src/sync/diagnostics/diagnosticExport.ts
export const DIAGNOSTIC_EXPORT_MAX_BYTES = 10 * 1024 * 1024;
export type EvidenceArea = "STATE_TRANSITION" | "FRONTIER" | "DISPOSITION" | "ROUTE" | "COMMIT_ACK" | "LIFECYCLE_GENERATION" | "RETRY_BACKOFF" | "PROVIDER_PROTECTION" | "CHECKPOINT_REHYDRATION" | "MIGRATION" | "COMPATIBILITY" | "SAFE_MODE" | "PERFORMANCE";
export interface DiagnosticEvent { readonly monotonicMillis: number; readonly area: EvidenceArea; readonly event: string; readonly fields: Readonly<Record<string, string>>; readonly deviceRef?: string; readonly operationRef?: string }
export interface DiagnosticSink { write(chunk: Uint8Array): Promise<void>; close(): Promise<void> }
export interface DiagnosticExportResult { readonly bytesWritten: number; readonly eventsWritten: number; readonly truncated: boolean; readonly cancelled: boolean }

const ALLOWED_FIELDS = new Set(["from", "to", "count", "sequence", "generation", "durationMs", "batchSize", "attempt", "delayMs", "outcome", "reasonCode", "routeKind", "sourceKind", "formatVersion"]);
const EVIDENCE_AREAS = new Set<EvidenceArea>(["STATE_TRANSITION", "FRONTIER", "DISPOSITION", "ROUTE", "COMMIT_ACK", "LIFECYCLE_GENERATION", "RETRY_BACKOFF", "PROVIDER_PROTECTION", "CHECKPOINT_REHYDRATION", "MIGRATION", "COMPATIBILITY", "SAFE_MODE", "PERFORMANCE"]);
const FORBIDDEN = /(key|secret|token|credential|capability|recovery|profile|photo|tag|history|payload|content|member)/i;
export function requireSafeDiagnostic(event: DiagnosticEvent): void {
  if (!Number.isSafeInteger(event.monotonicMillis) || event.monotonicMillis < 0 || !EVIDENCE_AREAS.has(event.area) || typeof event.event !== "string" || event.event.length < 1 || event.event.length > 80 || FORBIDDEN.test(event.event)) throw new Error("unsafe diagnostic event");
  if ((event.deviceRef !== undefined && (typeof event.deviceRef !== "string" || event.deviceRef.length > 200)) || (event.operationRef !== undefined && (typeof event.operationRef !== "string" || event.operationRef.length > 200))) throw new Error("unsafe diagnostic reference");
  const fields = Object.entries(event.fields); if (fields.length > 24) throw new Error("too many diagnostic fields");
  for (const [key, value] of fields) if (!ALLOWED_FIELDS.has(key) || value.length > 120 || FORBIDDEN.test(value)) throw new Error("Diagnostic field is not export-safe");
}

export async function streamDiagnosticExport(events: AsyncIterable<DiagnosticEvent>, sink: DiagnosticSink, signal?: AbortSignal): Promise<DiagnosticExportResult> {
  const encoder = new TextEncoder(); const devices = new Map<string, string>(); const operations = new Map<string, string>();
  let bytesWritten = 0; let eventsWritten = 0;
  for await (const event of events) {
    if (signal?.aborted === true) return { bytesWritten, eventsWritten, truncated: false, cancelled: true };
    requireSafeDiagnostic(event);
    const chunk = encoder.encode(`${JSON.stringify(exportRecord(event, devices, operations))}\n`);
    if (bytesWritten + chunk.length > DIAGNOSTIC_EXPORT_MAX_BYTES) { await sink.close(); return { bytesWritten, eventsWritten, truncated: true, cancelled: false }; }
    await sink.write(chunk); bytesWritten += chunk.length; eventsWritten++;
  }
  await sink.close(); return { bytesWritten, eventsWritten, truncated: false, cancelled: false };
}
function exportRecord(event: DiagnosticEvent, devices: Map<string, string>, operations: Map<string, string>): Record<string, unknown> {
  const output: Record<string, unknown> = { v: 1, monotonicMillis: event.monotonicMillis, area: event.area, event: event.event, fields: Object.fromEntries(Object.entries(event.fields).sort(([left], [right]) => left.localeCompare(right))) };
  if (event.deviceRef !== undefined) output["deviceAlias"] = alias(devices, event.deviceRef, "device");
  if (event.operationRef !== undefined) output["operationAlias"] = alias(operations, event.operationRef, "operation");
  return output;
}
function alias(values: Map<string, string>, raw: string, prefix: string): string { const current = values.get(raw); if (current !== undefined) return current; const value = `${prefix}-${values.size + 1}`; values.set(raw, value); return value; }
